query/src/sql/optimizer/heuristic/subquery_rewriter.rs

```diff
@@ -293,16 +306,44 @@ impl SubqueryRewriter {
                     subquery.data_type.clone()
                 };
 
-                let column_ref = ColumnBinding {
-                    database_name: None,
-                    table_name: None,
-                    column_name: name,
-                    index,
-                    data_type,
-                    visible_in_unqualified_wildcard: false,
+                let column_ref = Scalar::BoundColumnRef(BoundColumnRef {
+                    column: ColumnBinding {
+                        database_name: None,
+                        table_name: None,
+                        column_name: name,
+                        index,
+                        data_type,
+                        visible_in_unqualified_wildcard: false,
+                    },
+                });
+
+                let scalar = if flatten_info.from_count_func {
+                    // convert count aggregate function to multi_if function, if count() is null, then 0 else count()
```

Comment:
Why do we need to rewrite `COUNT(*)` into `CASE WHEN COUNT(*) IS NULL THEN 0 ELSE COUNT(*) END`?

AFAIK, `COUNT(*)` will never return a `NULL` value.

Reply:
Single join is similar to left outer join, if there isn't matched row on the right side, we'll add NULL value for the right side.  At the time, count(*) will be a NULL value.

Reply:
Ok, I get it.

So the `CASE WHEN` is actually applied to the column produced by single join  but not the `COUNT(*)`. 

And conincidentally, different from other set functions, `COUNT(*)` won't return NULL on empty input so it cannot utilize with the NULL produced by a single join as other aggregate functions.

Is that correct?

Reply:
Yes, you totally got it!